src/fromager/commands/build.py

```diff
@@ -560,6 +561,55 @@ def _nodes_to_string(nodes: typing.Iterable[dependency_graph.DependencyNode]) ->
     return ", ".join(sorted(node.key for node in nodes))
 
 
+def _calculate_optimal_max_workers(
+    graph: dependency_graph.DependencyGraph,
+    wkctx: context.WorkContext,
+    user_max_workers: int | None,
+) -> int:
+    """Calculate optimal max_workers based on graph parallelism.
+
+    Analyzes the dependency graph to determine the maximum number of packages
+    that can be built in parallel at any point. Uses this to optimize the
+    number of worker threads, avoiding wasteful allocation of idle workers.
+
+    Args:
+        graph: The dependency graph to analyze.
+        wkctx: The work context for topology calculation.
+        user_max_workers: User-specified max_workers, or None for automatic.
+
+    Returns:
+        The optimal number of workers to use.
+    """
+    # Analyze the topology to find maximum parallelism available
+    topo_for_analysis = graph.get_build_topology(context=wkctx)
+    max_parallelism = max(
+        (len(batch) for batch in topo_for_analysis.static_batches()), default=1
+    )
+
+    if user_max_workers is None:
+        # Use the smaller of CPU-based default and graph-based maximum
+        cpu_default = min(32, (os.cpu_count() or 1) + 4)
+        optimal_workers = min(cpu_default, max_parallelism)
+        logger.info(
+            "graph allows max %i parallel builds, using %i workers (cpu default: %i)",
```

Comment:
I don't know that I would understand this log message if I was reading it without having seen the code. Could you log each value separately with a short description before going through this if statement, then log the actual selected worker pool size being returned? Something like the messages below, for example.

You could add messages like "batch size from graph exceeds CPU count" or whatever to help with debugging, too.

```
"batch size from graph: %d"
"CPU count: %d"
"CPU pool size: %d"
"user requested pool size: %d"
"worker pool size: %d"
```